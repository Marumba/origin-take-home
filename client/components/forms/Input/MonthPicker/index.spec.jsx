import React from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen, fireEvent } from '~/helpers/testing.customRtl';

import InputMonthPicker from './';

const todayMonth = 1;
const initialDate = '2021-2';
const initialYear = 2021;
const initialState = {
	month: 'February',
	monthPointer: 1,
	monthsTotal: 1,
	year: initialYear
};

jest.mock('~/helpers/utils', () => ({
	getYearMonthDatetime: jest.fn().mockImplementationOnce(() => ({
		month: todayMonth,
		year: initialYear,
		datetime: 12312321
	}))
}));

describe('Input Month Picker', () => {
	const marchState = {
		month: 'March',
		monthPointer: 2,
		monthsTotal: 2,
		year: 2021
	};
	const aprilState = {
		month: 'April',
		monthPointer: 3,
		monthsTotal: 3,
		year: 2021
	};
	const onChangeMock = jest.fn();

	it('should render', () => {
		const { container } = render(<InputMonthPicker id="testeId" name="testeId" />);
		expect(screen.getByRole('textbox')).toBeInTheDocument();
		expect(container.firstChild).toMatchSnapshot();
	});

	describe('Changing dates', () => {
		it('should allow only future months', () => {
			render(<InputMonthPicker id="testeId" name="testeId" onChange={onChangeMock} />);
			expect(onChangeMock).toHaveBeenCalledWith(initialDate, initialState.monthsTotal, initialState);
			const leftArrow = screen.queryAllByRole('button')[0];

			userEvent.click(leftArrow);
			expect(onChangeMock).toHaveBeenCalledWith(initialDate, initialState.monthsTotal, initialState);
		});

		it('should go up and down, month by month, when clicking on the arrows', () => {
			render(<InputMonthPicker id="testeId" name="testeId" onChange={onChangeMock} />);
			const [leftArrow, rightArrow] = screen.getAllByRole('button');

			userEvent.click(rightArrow);
			expect(onChangeMock).toHaveBeenCalledWith('2021-3', 2, marchState);

			userEvent.click(rightArrow);
			expect(onChangeMock).toHaveBeenCalledWith('2021-4', 3, aprilState);

			userEvent.click(leftArrow);
			expect(onChangeMock).toHaveBeenCalledWith('2021-3', 2, marchState);
		});
		it('should have the same behavior of the arrow buttons when navigating with the keyboard arrows', () => {
			render(<InputMonthPicker id="testeId" name="testeId" onChange={onChangeMock} />);
			const [leftArrow, rightArrow] = screen.getAllByRole('button');

			userEvent.tab();
			expect(leftArrow).toHaveFocus();

			userEvent.tab();
			userEvent.tab();
			expect(rightArrow).toHaveFocus();
			fireEvent.keyUp(rightArrow, { key: 'ArrowRight', which: 38, keyCode: 38 });
			fireEvent.keyDown(rightArrow, { key: 'ArrowRight', which: 38, keyCode: 38 });
			expect(rightArrow).toHaveStyleRule('transform', 'rotate(180deg) scale(1.4)', { target: 'svg' });
			expect(onChangeMock).toHaveBeenCalledWith('2021-3', 2, marchState);

			fireEvent.keyPress(rightArrow, { key: 'ArrowRight', which: 38, keyCode: 38 });
			expect(onChangeMock).toHaveBeenCalledWith('2021-4', 3, aprilState);

			userEvent.tab();
			userEvent.tab();
			expect(leftArrow).toHaveFocus();
			fireEvent.keyUp(leftArrow, { key: 'ArrowLeft', which: 37, keyCode: 37 });
			expect(onChangeMock).toHaveBeenCalledWith('2021-3', 2, marchState);
			expect(leftArrow).toHaveStyleRule('transform', 'scale(1.4)', { target: 'svg' });
		});
	});

	// describe('')
});
