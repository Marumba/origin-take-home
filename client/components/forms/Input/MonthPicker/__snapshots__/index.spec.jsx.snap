// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Month Picker should render 1`] = `
.emotion-0:focus {
  background: red;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-2 input {
  width: auto;
  -webkit-transform: translateX(-50%);
  -moz-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  position: absolute;
  left: 50%;
  z-index: 1;
}

.emotion-4 {
  font-family: 'Work Sans',sans-serif;
  font-weight: 400;
  font-size: 12px;
  line-height: 150%;
  margin-bottom: 4px;
  color: #1E2A32;
}

@media screen and (min-width: 992px) {
  .emotion-4 {
    font-size: 14px;
  }
}

.emotion-6 {
  border: 1px solid #E9EEF2;
  border-radius: 4px;
  background: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 56px;
  position: relative;
  padding: 0 14px;
}

.emotion-6 svg+input {
  padding-left: 7px;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  width: 50%;
  height: 56px;
  padding: 0 14px;
  position: absolute;
  z-index: 10;
  -webkit-transition: -webkit-transform 100ms ease-in-out;
  transition: transform 100ms ease-in-out;
  left: 0;
}

@media screen and (min-width: 992px) {
  .emotion-8 {
    padding: 0;
  }

  .emotion-8 svg {
    margin-left: 12px;
  }

  .emotion-8:hover svg {
    -webkit-transform: scale(1.2);
    -moz-transform: scale(1.2);
    -ms-transform: scale(1.2);
    transform: scale(1.2);
  }
}

.emotion-8:active svg {
  -webkit-transform: scale(1.4);
  -moz-transform: scale(1.4);
  -ms-transform: scale(1.4);
  transform: scale(1.4);
  -webkit-transition: 0.1s;
  transition: 0.1s;
}

.emotion-10 {
  border: 1px solid #E9EEF2;
  border-radius: 4px;
  background: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 56px;
  border: 0;
  cursor: pointer;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  z-index: 5;
}

.emotion-12 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-family: 'Work Sans',sans-serif;
  font-weight: 400;
  font-size: 14px;
  line-height: 150%;
}

@media screen and (min-width: 992px) {
  .emotion-12 {
    font-size: 16px;
  }
}

.emotion-12.semibold,
.emotion-12 strong {
  font-weight: 600;
}

.emotion-12.highEmphasis {
  color: #1E2A32;
}

.emotion-12.lowEmphasis {
  color: #708797;
}

.emotion-14 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-family: 'Work Sans',sans-serif;
  font-weight: 400;
  font-size: 14px;
  line-height: 150%;
}

@media screen and (min-width: 992px) {
  .emotion-14 {
    font-size: 16px;
  }
}

.emotion-14.semibold,
.emotion-14 strong {
  font-weight: 600;
}

.emotion-14.highEmphasis {
  color: #1E2A32;
}

.emotion-14.lowEmphasis {
  color: #708797;
}

.emotion-16 {
  font-family: 'Rubik',sans-serif;
  font-weight: 500;
  font-size: 20px;
  line-height: 120%;
  color: #4D6475;
  padding: 14px 0;
  width: 100%;
}

@media screen and (min-width: 992px) {
  .emotion-16 {
    font-size: 24px;
  }
}

.emotion-16.highEmphasis {
  color: #1E2A32;
}

.emotion-18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  width: 50%;
  height: 56px;
  padding: 0 14px;
  position: absolute;
  z-index: 10;
  -webkit-transition: -webkit-transform 100ms ease-in-out;
  transition: transform 100ms ease-in-out;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  right: 0;
}

@media screen and (min-width: 992px) {
  .emotion-18 {
    padding: 0;
  }

  .emotion-18 svg {
    margin-left: 12px;
  }

  .emotion-18:hover svg {
    -webkit-transform: scale(1.2);
    -moz-transform: scale(1.2);
    -ms-transform: scale(1.2);
    transform: scale(1.2);
  }
}

.emotion-18:active svg {
  -webkit-transform: scale(1.4);
  -moz-transform: scale(1.4);
  -ms-transform: scale(1.4);
  transform: scale(1.4);
  -webkit-transition: 0.1s;
  transition: 0.1s;
}

@media screen and (min-width: 992px) {
  .emotion-18 svg {
    margin-left: 0;
    margin-right: 12px;
  }

  .emotion-18:hover svg {
    -webkit-transform: rotate(180deg) scale(1.2);
    -moz-transform: rotate(180deg) scale(1.2);
    -ms-transform: rotate(180deg) scale(1.2);
    transform: rotate(180deg) scale(1.2);
  }
}

.emotion-18:active svg {
  -webkit-transform: rotate(180deg) scale(1.4);
  -moz-transform: rotate(180deg) scale(1.4);
  -ms-transform: rotate(180deg) scale(1.4);
  transform: rotate(180deg) scale(1.4);
}

<div
  class="emotion-0 emotion-1"
>
  <div
    class="emotion-2 emotion-3"
  >
    <label
      class="emotion-4 emotion-5"
      for="testeId"
    />
    <div
      class="emotion-6 emotion-7"
    >
      <button
        class="emotion-8 emotion-9"
        type="button"
      >
        <div />
      </button>
      <div
        class="emotion-10 emotion-11"
      >
        <div
          class="highEmphasis semibold emotion-12 emotion-13"
        >
          February
        </div>
        <div
          class="highEmphasis emotion-14 emotion-15"
        >
          2021
        </div>
      </div>
      <input
        class="emotion-16 emotion-17"
        id="testeId"
        name="testeId"
        readonly=""
        type="text"
        value="2021-2"
      />
      <button
        class="emotion-18 emotion-19"
        type="button"
      >
        <div
          transform="rotate(180)"
        />
      </button>
    </div>
  </div>
</div>
`;
